eth_utils/abi.py

```diff
@@ -1,68 +1,784 @@
+
+    abi_type = str(abi_element.get("type", ""))
+    if abi_type == "fallback" or abi_type == "receive":
+        return signature.format(name=abi_type, input_types="")
+
+    if abi_type == "constructor":
+        fn_name = abi_type
+    else:
+        fn_name = str(abi_element.get("name", abi_type))
+
+    return signature.format(
+        name=fn_name, input_types=",".join(get_abi_input_types(abi_element))
+    )
+
+
+def filter_abi_by_name(name: str, contract_abi: ABI) -> Sequence[ABIElement]:
+    """
+    Get one or more function and event ABIs by name.
+
+    :param name: Name of the function, event or error.
+    :type name: `str`
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: Function or event ABIs with matching name.
+    :rtype: `Sequence[ABIElement]`
+
+    .. doctest::
+
+            >>> from eth_utils.abi import filter_abi_by_name
+            >>> abi = [
+            ...     {
+            ...         "constant": False,
+            ...         "inputs": [],
+            ...         "name": "func_1",
+            ...         "outputs": [],
+            ...         "type": "function",
+            ...     },
+            ...     {
+            ...         "constant": False,
+            ...         "inputs": [
+            ...             {"name": "a", "type": "uint256"},
+            ...         ],
+            ...         "name": "func_2",
+            ...         "outputs": [],
+            ...         "type": "function",
+            ...     },
+            ...     {
+            ...         "constant": False,
+            ...         "inputs": [
+            ...             {"name": "a", "type": "uint256"},
+            ...             {"name": "b", "type": "uint256"},
+            ...         ],
+            ...         "name": "func_3",
+            ...         "outputs": [],
+            ...         "type": "function",
+            ...     },
+            ...     {
+            ...         "constant": False,
+            ...         "inputs": [
+            ...             {"name": "a", "type": "uint256"},
+            ...             {"name": "b", "type": "uint256"},
+            ...             {"name": "c", "type": "uint256"},
+            ...         ],
+            ...         "name": "func_4",
+            ...         "outputs": [],
+            ...         "type": "function",
+            ...     },
+            ... ]
+            >>> filter_abi_by_name("func_1", abi)
+            [{'constant': False, 'inputs': [], 'name': 'func_1', 'outputs': [], \
+'type': 'function'}]
+    """
+    return [
+        abi
+        for abi in contract_abi
+        if (
+            (
+                abi["type"] == "function"
+                or abi["type"] == "event"
+                or abi["type"] == "error"
+            )
+            and abi["name"] == name
+        )
+    ]
+
+
+def filter_abi_by_type(type: str, contract_abi: ABI) -> Sequence[ABIElement]:
```

Comment:
`type` is a reserved word in Python. Let's change this to `abi_type` and we should maybe change `name` to `abi_name` also for consistency.

```suggestion
def filter_abi_by_type(abi_type: str, contract_abi: ABI) -> Sequence[ABIElement]:
```